File: gasm/src/GAsm.cpp
//
// Created by mateu on 14.11.2025.
//

#include <filesystem>
#include <fstream>
#include "utils.h"
#include "../include/GAsm.h"
#include <nlohmann/json.hpp>
#include <random>
#include <iostream>
#include <thread>
#include <cfloat>
#include <boost/multiprecision/cpp_bin_float.hpp>

GAsm::GAsm() : runner_(1), population_(0), fitness_(1), rank_(1) {
    unsigned int numThreads = std::thread::hardware_concurrency();
    if (numThreads == 0) numThreads = 4; // fallback
    runners_.reserve(numThreads);

    for (unsigned int i = 0; i < numThreads; i++) {
        runners_.emplace_back(std::move(Runner()));
    }
}

GAsm::GAsm(const std::string &filename) : runner_(1) {
    unsigned int numThreads = std::thread::hardware_concurrency();
    if (numThreads == 0) numThreads = 4; // fallback
    runners_.reserve(numThreads);

    for (unsigned int i = 0; i < numThreads; i++) {
        runners_.emplace_back(std::move(Runner()));
    }
    using nlohmann::json;
    // Read file
    std::ifstream file(filename, std::ios::binary);
    if (!file) throw std::runtime_error("Cannot open JSON file: " + filename);

    std::string content((std::istreambuf_iterator<char>(file)),
                        std::istreambuf_iterator<char>());
    file.close();

    json j = json::parse(content);

    // Load primitive fields
    populationSize      = j["populationSize"];
    individualMaxSize   = j["individualMaxSize"];
    mutationProbability = j["mutationProbability"];
    crossoverProbability= j["crossoverProbability"];
    maxGenerations      = j["maxGenerations"];
    goalFitness         = j["goalFitness"];
    maxProcessTime      = j["maxProcessTime"];
    outputFolder        = j["outputFolder"];
    checkPointInterval  = j["checkPointInterval"];

    // Load register length
    setRegisterLength(j["registerLength"]);

    // Load inputs & targets
    inputs  = j["inputs"].get<std::vector<std::vector<double>>>();
    targets = j["targets"].get<std::vector<std::vector<double>>>();

    // Load fitness & rank
    fitness_ = j["fitness"].get<std::vector<double>>();
    rank_    = j["rank"].get<std::vector<double>>();

    // Load best individual
    std::string ascii = j["bestIndividual"].get<std::string>();
    size_t len = ascii.size();
    uint8_t *bytecode = GAsmParser::ascii2Bytecode(ascii, len);
    std::vector<uint8_t> individual(bytecode, bytecode + len);
    delete[] bytecode;
    bestIndividual = std::move(individual);

    // Deserialize population
    population_.clear();
    for (auto &ascii_value : j["population"])
    {
        ascii = ascii_value.get<std::string>();

        len = ascii.size();
        bytecode = GAsmParser::ascii2Bytecode(ascii, len);

        individual = std::vector<uint8_t>(bytecode, bytecode + len);
        delete[] bytecode;

        population_.push_back(std::move(individual));
    }

    // Load history
    if (j.contains("history"))
        hist = Hist(j["history"]);
    else
        hist = Hist();   // empty history
}

GAsm::~GAsm() = default;

nlohmann::json GAsm::toJson() {
    using nlohmann::json;
    json j;
    // Save simple fields
    j["populationSize"]    = populationSize;
    j["individualMaxSize"] = individualMaxSize;
    j["mutationProbability"] = mutationProbability;
    j["crossoverProbability"] = crossoverProbability;
    j["maxGenerations"] = maxGenerations;
    j["goalFitness"] = goalFitness;
    j["outputFolder"] = outputFolder;
    j["checkPointInterval"] = checkPointInterval;
    j["maxProcessTime"] = maxProcessTime;

    // Save inputs
    j["inputs"] = inputs;
    j["targets"] = targets;

    // Save bestIndividual
    std::string ascii = GAsmParser::bytecode2Ascii(
            bestIndividual.data(),
            bestIndividual.size()
    );
    j["bestIndividual"] = ascii;

    // Save population (as ASCII)
    j["population"] = json::array();
    for (const auto &individual : population_)
    {
        ascii = GAsmParser::bytecode2Ascii(
                individual.data(),
                individual.size()
        );

        j["population"].push_back(ascii);
    }

    // Save fitness & rank
    j["fitness"] = fitness_;
    j["rank"]    = rank_;

    // Save history
    j["history"] = hist.toJson();

    // Save register length
    j["registerLength"] = runner_.getRegisterLength();

    return j;
}

void GAsm::save2File(const std::string& filename) {
    // Write to file
    std::ofstream f(filename);
    if (!f.is_open()) {
        std::cerr << "FAILED to open file: " << filename << "\n";
        return;
    }
    std::cout << "Writing to file: " << std::filesystem::absolute(filename) << std::endl;
    f << toJson().dump(4);   // pretty print
    f.close();
    std::cout << "File written successfully!" << std::endl;
}

void GAsm::printHeader(const GAsm* const self) {
    std::cout << "-- GAsm evolution --" << std::endl;
    std::cout << "Max individual size: " << self->individualMaxSize << std::endl;
    std::cout << "Population size: " << self->populationSize << std::endl;
    std::cout << "Crossover probability: " << self->crossoverProbability << std::endl;
    std::cout << "Mutation probability: " << self->mutationProbability << std::endl;
    std::cout << "Max generations: " << self->maxGenerations << std::endl;
    std::cout << "Register length: " << self->runner_.getRegisterLength() << std::endl;
    std::cout << "Max process time: " << self->maxProcessTime << std::endl;
    std::cout << "Minimize: " << (self->minimize ? "True" : "False") << std::endl;
    std::cout << "Output folder: " << self->outputFolder << std::endl;
    std::cout << "Checkpoint interval: " << self->checkPointInterval << std::endl;
    std::cout << "Goal fitness: " << self->goalFitness << std::endl;
    std::cout << "NaN penalty: " << self->nanPenalty << std::endl;
    std::cout << "Number of cores: " << self->runners_.size() << std::endl;
    std::cout << "----------------------------------" << std::endl;
}

double GAsm::printGenerationStats(int generation) {
    boost::multiprecision::cpp_bin_float_quad avgFitness = 0.0;
    double bestFitness = minimize ? DBL_MAX: -DBL_MAX; // NOLINT
    double avgSize = 0.0;
    size_t bestIndividualIndex = 0;
    for (int i = 0; i < population_.size(); i++) {
        avgFitness += std::isfinite(fitness_[i]) ? fitness_[i] : 0;
        avgSize += (double)population_[i].size();
        if (minimize ? fitness_[i] < bestFitness : fitness_[i] > bestFitness) {
            bestFitness = fitness_[i];
            bestIndividualIndex = i;
        }
    }
    avgFitness /= (double)population_.size();
    avgSize /= (double)population_.size();
    bestIndividual = population_[bestIndividualIndex];
    auto convertedAvgFitness = (double)avgFitness;
    hist.add(generation, bestFitness, convertedAvgFitness, avgSize, bestIndividual);
    std::cout << "Generation: " << generation
              << ", Avg Fitness: " << std::scientific << convertedAvgFitness
              << ", Best Fitness: " << std::fixed << std::setprecision(2) << bestFitness
              << ", Avg Size: " << avgSize << std::endl;
    return bestFitness;
}

void GAsm::parallelEvolve(const std::vector<std::vector<double>>& inputs_,
                          const std::vector<std::vector<double>>& targets_)
{
    using namespace std::chrono;
    // TODO checkpoints
    // TODO from checkpoint
    this->inputs = inputs_;
    this->targets = targets_;

    // Resize population and fitness
    population_.resize(populationSize);
    fitness_.resize(populationSize);
    rank_.resize(populationSize);
    individualMutexes_.resize(populationSize);
    for (auto& m : individualMutexes_)
        m = std::make_unique<std::mutex>();
    for (auto& ind : population_) {
        ind.reserve(individualMaxSize);
    }
    size_t numThreads = runners_.size();
    std::vector<std::thread> threads;
    threads.reserve(numThreads);
    size_t chunk = (populationSize + numThreads - 1) / numThreads;

    printHeader(this);

    std::cout << "Initializing population" << std::endl;
    for (size_t t = 0; t < numThreads; ++t) {
        size_t start = t * chunk;
        size_t end = std::min<size_t>(start + chunk, populationSize);

        threads.emplace_back([&, t, start, end]() {
            // each runner gets its chunk and works
            runners_[t].dispatchGrow(this, start, end, t == 0);  // set first to verbose
        });
    }

    for (auto &th : threads) th.join(); // wait for all the threads
    threads.clear();
    std::cout << std::endl;
    printGenerationStats(0);

    auto evolutionStart = high_resolution_clock::now();

    for (int generation = 0; generation < maxGenerations; generation++) {
        for (size_t t = 0; t < numThreads; ++t) {
            size_t start = t * chunk;
            size_t end = std::min<size_t>(start + chunk, populationSize);

            threads.emplace_back([&, t, start, end]() {
                // each runner gets its chunk and works
                runners_[t].dispatchEvolve(this, start, end, t == 0);  // set first to verbose
            });
        }
        for (auto &th : threads) th.join(); // wait for all the threads
        threads.clear();
        std::cout << std::endl;

        double fitness = printGenerationStats(generation + 1);

        // Early stopping
        if (minimize) {
            if (fitness <= goalFitness) break;
        } else {
            if (fitness >= goalFitness) break;
        }
    }
    double elapsed = duration<double>(high_resolution_clock::now() - evolutionStart).count();
    std::cout << "Evolution finished, took: ";
    printTime(elapsed);
    std::cout << std::endl;
}

void GAsm::evolve(const std::vector<std::vector<double>>& inputs_,
                  const std::vector<std::vector<double>>& targets_)
{
    using namespace std::chrono;
    // TODO checkpoints
    // TODO from checkpoint
    this->inputs = inputs_;
    this->targets = targets_;

    // Resize population and fitness
    population_.resize(populationSize);
    fitness_.resize(populationSize);
    rank_.resize(populationSize);
    individualMutexes_.resize(populationSize);
    for (auto& m : individualMutexes_)
        m = std::make_unique<std::mutex>();
    for (auto& ind : population_) {
        ind.reserve(individualMaxSize);
    }

    printHeader(this);

    std::cout << "Initializing population" << std::endl;
    auto initStart = high_resolution_clock::now();
    for (size_t i = 0; i < populationSize; i++) {
        int progress = ((int)i + 1) * 100 / (int) populationSize;
        double elapsed = duration<double>(high_resolution_clock::now() - initStart).count();
        printProgressBar(progress, elapsed);
        (*growFunction_)(this, population_[i]);
//        std::cout << std::endl << GAsmParser::bytecode2Text(population_[i].data(), population_[i].size()) << std::endl;
        std::pair<double, double> fitRank = (*fitnessFunction_)(this, this->runner_, population_[i]);
//        std::cout << "Fitness: " << fitRank.first << std::endl;
//        std::cout << "Rank: " << fitRank.second << std::endl;
//        std::cout << "Individual: " << GAsmParser::bytecode2Text(population_[i].data(), population_[i].size()) << std::endl;
        fitness_[i] = fitRank.first;
        rank_[i] = fitRank.second;
    }
    std::cout << std::endl;
    printGenerationStats(0);

    static thread_local std::mt19937 engine(std::random_device{}());
    std::uniform_real_distribution<double> dist(0, 1);

    auto evolutionStart = high_resolution_clock::now();

    for (int generation = 0; generation < maxGenerations; generation++) {
        auto genStart = high_resolution_clock::now();
        for (int i = 0; i < populationSize; i++) {
            selectionFunction_->selectMinimal = !minimize; // worst is not minimized
            size_t worstIndex = (*selectionFunction_)(this);
            selectionFunction_->selectMinimal = minimize;  // best is minimized
            if (dist(engine) < crossoverProbability) {
                size_t bestIndex1 = (*selectionFunction_)(this);
                size_t bestIndex2 = (*selectionFunction_)(this);

                (*crossoverFunction_)(this, population_[worstIndex], population_[bestIndex1], population_[bestIndex2]);
            } else {
                size_t bestIndex = (*selectionFunction_)(this);
                (*mutationFunction_)(this, population_[worstIndex], population_[bestIndex]);
            }

            std::pair<double, double> fitRank = (*fitnessFunction_)(this, this->runner_, population_[worstIndex]);
            fitness_[worstIndex] = fitRank.first;
            rank_[worstIndex] = fitRank.second;
            int progress = (i + 1) * 100 / (int) populationSize;
            double elapsed = duration<double>(high_resolution_clock::now() - genStart).count();
            printProgressBar(progress, elapsed);
        }
        std::cout << std::endl;

        double fitness = printGenerationStats(generation + 1);

        // Early stopping
        if (minimize) {
            if (fitness <= goalFitness) break;
        } else {
            if (fitness >= goalFitness) break;
        }
    }
    double elapsed = duration<double>(high_resolution_clock::now() - evolutionStart).count();
    std::cout << "Evolution finished, took: ";
    printTime(elapsed);
    std::cout << std::endl;
}

void GAsm::setProgram(const std::vector<uint8_t>& program) {
    runner_.setProgram(program);
}

size_t GAsm::run(std::vector<double>& inputs_) {
    return runner_.run(inputs_, maxProcessTime);
}

double GAsm::runAll(const std::vector<uint8_t>& program, std::vector<std::vector<double>>& outputs) {
    runner_.setProgram(program);
    double sumTime = 0.0;
    for (const auto& input : inputs) {
        std::vector<double> output = input;
        sumTime += (double)runner_.run(output, maxProcessTime);
        outputs.push_back(std::move(output));
    }
    return sumTime;
}

void GAsm::makeCheckpoint() {
    // TODO make a checkpoint
}

File: gasm/python/gasm.py
class GAsm:
    """
    Genetic Assembly Machine (GAsm)

    Main engine for genetic programming.
    Provides control over population, mutation, crossover, growth,
    fitness evaluation, and evolution history.

    ---------------------------------------------------------------------
    Attributes
    ---------------------------------------------------------------------

    populationSize : int
        Number of individuals in the population.

    individualMaxSize : int
        Maximum allowed number of instructions in an individual.

    mutationProbability : float
        Probability of mutation during evolution.

    crossoverProbability : float
        Probability of crossover during evolution.

    maxGenerations : int
        Maximum number of generations to evolve.

    goalFitness : float
        Desired fitness target at which evolution stops.

    minimize : bool
        If True → lower fitness is better (minimization mode).
        If False → higher fitness is better.

    registerLength : int
        Number of registers available during program execution.

    maxProcessTime : int
        Maximum number of instructions executed before halting.

    best : Individual (read-only)
        Returns the best individual discovered so far.

    hist : Hist (read-only)
        Full history of evolution.
        Allows indexing like `hist[i]` to obtain an Entry.

    outputFolder : str
        Folder where logs, checkpoints, and outputs are stored.

    nanPenalty : float
        Fitness penalty applied if the program outputs NaN.

    useCompile : bool
        Enables or disables JIT compilation (if supported).

    checkpointInterval : int
        Save a checkpoint every N generations (0 = disabled).

    ---------------------------------------------------------------------
    Methods
    ---------------------------------------------------------------------
    """

    # ------------------------------------------------------------------
    # Core execution
    # ------------------------------------------------------------------

    def run(self, inputs: list[int]) -> list[int]:
        """Execute the current program on one input vector."""

    def runAll(self, inputs: list[list[int]]) -> list[list[int]]:
        """Execute the program on all inputs."""

    def setProgram(self, program: list[str]) -> None:
        """Replace the current program with one supplied by the user."""

    # ------------------------------------------------------------------
    # Evolution
    # ------------------------------------------------------------------

    def evolve(self, inputs: list[list[int]], outputs: list[list[int]]) -> None:
        """
        Run full evolutionary process.

        Parameters
        ----------
        inputs : list[list[int]]
            Training inputs.
        outputs : list[list[int]]
            Expected outputs.

        Evolution stops when:
            - goalFitness is reached OR
            - maxGenerations is exceeded.
        """

    # ------------------------------------------------------------------
    # Serialization
    # ------------------------------------------------------------------

    def save2File(self, path: str) -> None:
        """Save engine state (population + config + history) to a JSON file."""

    @staticmethod
    def fromJson(path: str) -> "GAsm":
        """Load engine state from a JSON file and return a new GAsm instance."""

    # ------------------------------------------------------------------
    # Selection
    # ------------------------------------------------------------------

    def setSelection(self, mode: str, param: int | float | None = None) -> None:
        """
        Set selection method.

        Parameters
        ----------
        mode : Literal["Tournament", "Truncation", "Boltzman", "Rank", "Roulette"]
            Selection strategy.

        param :
            Additional parameter depending on mode:
             - Tournament: tournament size (int)
             - Truncation: top-k percentage or count
             - Boltzman: temperature parameter
             - Rank: unused or tuning parameter
             - Roulette: no parameter required
        """

    # ------------------------------------------------------------------
    # Grow policy
    # ------------------------------------------------------------------

    def setGrow(self, mode: str, param: int | None = None) -> None:
        """
        Set growth mode for generating trees.

        Parameters
        ----------
        mode : Literal["Size", "Full", "Tree"]
            Grow strategy.

        param :
            Size limit or depth depending on mode.
        """

    # ------------------------------------------------------------------
    # Mutation & Crossover
    # ------------------------------------------------------------------

    def setMutation(self, mode: str) -> None:
        """
        Set mutation type.

        Parameters
        ----------
        mode : Literal["Hard", "Soft"]
        """

    def setCrossover(self, mode: str) -> None:
        """
        Set crossover method.

        Parameters
        ----------
        mode : Literal["OnePoint", "TwoPoint", "TwoPointSize", "Uniform"]
        """

    # ------------------------------------------------------------------
    # Generators (CNG & RNG)
    # ------------------------------------------------------------------

    def set_cng(self, type: str, start: float) -> None:
        """
        Set Custom Number Generator (CNG).

        Parameters
        ----------
        type : Literal["increment", "constant"]
        start : float
        """

    def set_rng(self, type: str, start: float) -> None:
        """
        Set Random Number Generator (RNG).

        Parameters
        ----------
        type : Literal["increment", "constant"]
        start : float
        """


class Hist:
    """
    Evolution history container (read-only).

    Behaves like a Python sequence:
        len(hist), hist[i], iteration, slicing

    Each entry is an `Entry` describing one generation.

    Methods
    -------
    __len__()
        Number of entries.
    __getitem__(i)
        Get the Entry at index i.
    """

    def __len__(self) -> int:
        """Return number of generations in the history."""
        ...

    def __getitem__(self, idx: int) -> "Entry":
        """
        Get history entry by index.

        Parameters
        ----------
        idx : int
            Index of the generation.

        Returns
        -------
        Entry
            Statistics and best individual of that generation.
        """
        ...


class Entry:
    """
    Statistics for a single generation.

    Attributes
    ----------
    generation : int
        Generation number.

    bestFitness : float
        Fitness of the best individual.

    avgFitness : float
        Average population fitness.

    avgSize : float
        Average program size.

    bestIndividual : Individual
        Best Individual of the generation.
    """

    @property
    def generation(self) -> int:
        """Generation index."""
        ...

    @property
    def bestFitness(self) -> float:
        """Fitness of the best program."""
        ...

    @property
    def avgFitness(self) -> float:
        """Average fitness of the population."""
        ...

    @property
    def avgSize(self) -> float:
        """Average program size."""
        ...

    @property
    def best(self) -> "Individual":
        """Best program (Individual) from this generation."""
        ...


class Individual:
    """
    Evolvable program represented as bytecode.

    Methods
    -------
    run(inputs)
        Executes the program on given input registers.
    toString()
        Return a human-readable text version of the bytecode.

    Attributes
    ----------
    maxProcessTime : int
        Maximum number of instructions.

    registerLength : int
        Number of registers used.

    cng : Callable[[], float]
        Constant number generator (CNG) used for growth/mutation.

    rng : Callable[[], float]
        Random number generator (RNG) used for growth/mutation.

    useCompile : bool
        Whether compiled execution is used internally.
    """

    def run(self, inputs: list[float]) -> int:
        """
        Execute the program.

        Parameters
        ----------
        inputs : list[float]
            Initial register values.

        Returns
        -------
        int
            Execution result (instruction count or program output, depending on implementation).
        """
        ...

    def toString(self) -> str:
        """
        Return the textual representation of the bytecode.

        Returns
        -------
        str
            Pretty-printed assembly-like representation.
        """
        ...


